src/Workspaces/Core/Portable/Remote/RemoteHostSessions.cs

```diff
@@ -0,0 +1,280 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Execution;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// Helper type of common session case
+    /// 
+    /// RemoteHostClient.Session with 1 solution
+    /// </summary>
+    internal sealed class SessionWithSolution : IDisposable
+    {
+        private readonly RemoteHostClient.Session _session;
+        private readonly PinnedRemotableDataScope _scope;
+        private readonly CancellationToken _cancellationToken;
+
+        public static async Task<SessionWithSolution> CreateAsync(RemoteHostClient.Session session, PinnedRemotableDataScope scope, CancellationToken cancellationToken)
+        {
+            var sessionWithSolution = new SessionWithSolution(session, scope, cancellationToken);
+
+            try
+            {
+                await session.RegisterPinnedRemotableDataScopeAsync(scope).ConfigureAwait(false);
+                return sessionWithSolution;
+            }
+            catch
```

Comment:
all kind of exceptions due to us disconnecting connection when cancellation is thrown. I changed it to actually throw on cancellation.